okreplay-core/src/main/java/okreplay/Optional.java

```diff
@@ -0,0 +1,184 @@
+package okreplay;
+
+import java.io.Serializable;
+import java.util.Set;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import static okreplay.Util.checkNotNull;
+
+abstract class Optional<T> implements Serializable {
```

Comment:
Maybe use a library for this? It could be easier: https://github.com/memoizr/retro-optional.